Check that sparse matrix is valid before constructing a CHOLMOD.Sparse

andreasnoack opened the pull request:
These are just some consistency checks of the inputs before constructing a `CHOLMOD.Sparse`. Julia's `SparseMatrixCSC` should always pass the tests unless the buffers have been directly modified. However, as seen in JuliaLang/LinearAlgebra.jl#394, serialization might corrupt the sparse matrices which can lead to cryptic bugs. The checks introduced in this PR would probably have made it easier to detect the issue.

Fixes JuliaLang/LinearAlgebra.jl#394

andreasnoack committed:
Check that sparse matrix is valid before constructing a CHOLMOD.Sparse

Don't use length of underlying arrays to query sizes

Rename variables in zero based constructors

Sacha0 commented on base/sparse/cholmod.jl:
```diff
@@ -860,7 +860,18 @@ convert(::Type{Dense}, A::Sparse) = sparse_to_dense(A)
-    # check if columns are sorted
+    # checks
+    ## length of input
+    if length(nzval) != length(rowval)
```
In general CSC does not require that `length(nzval) == length(rowval)`?

andreasnoack replied:
I can see why you'd like to allow `length(nzval) > colptr[end]` but why would you allow `length(nzval) != length(rowval)`? It also appears that CHOLMOD assumes that these have the same size.

Sacha0 commented on base/sparse/cholmod.jl:
```diff
@@ -860,7 +860,18 @@ convert(::Type{Dense}, A::Sparse) = sparse_to_dense(A)
+    if length(nzval) != length(rowval)
+        throw(ArgumentError("nzval and rowval must have same length"))
+    end
+    if length(colptr) != n + 1
```
Perhaps `length(colptr) >= n + 1`?

andreasnoack replied:
Why shouldn't `length(colptr) == n + 1`? Again, that seems to be the assumption of CHOLMOD.

Sacha0 commented on base/sparse/cholmod.jl:
```diff
@@ -860,7 +860,18 @@ convert(::Type{Dense}, A::Sparse) = sparse_to_dense(A)
+    if length(colptr) != n + 1
+        throw(ArgumentError("length of colptr must be n + 1 = $(n + 1) but was $(length(colptr))"))
+    end
+    if colptr[end] != length(rowval)
```
Like above, CSC does not require that `length(rowval) == colptr[n + 1]`, and we do not generally enforce that condition (nor should we)? Perhaps `length(rowval) >= colptr[n + 1]` and likewise for `nzval`?

andreasnoack replied:
I can see why you don't want this so I'll change it. However, I think I might have assumed this all over the CHOLMOD wrappers so there might be other places where the number of non-zeros is queried with `length(nzval)` which is then wrong.

Sacha0 commented on base/sparse/cholmod.jl:
```diff
@@ -898,6 +909,17 @@ function (::Type{Sparse}){Tv<:VTypes}(m::Integer, n::Integer,
 
 function (::Type{Sparse}){Tv<:VTypes}(A::SparseMatrixCSC{Tv,SuiteSparse_long}, stype::Integer)
+    ## Check length of input. This should never fail but see #20024
+    if length(A.nzval) != length(A.rowval)
```
https://github.com/JuliaLang/julia/pull/20464/files#r99494647 ?

Sacha0 commented on base/sparse/cholmod.jl:
```diff
@@ -898,6 +909,17 @@ function (::Type{Sparse}){Tv<:VTypes}(m::Integer, n::Integer,
+    if length(A.colptr) != size(A,2) + 1
+        throw(ArgumentError("length of colptr must be size(A,2) + 1 = $(size(A,2) + 1) but was $(length(A.colptr))"))
+    end
+    if A.colptr[end] != length(A.rowval) + 1
```
https://github.com/JuliaLang/julia/pull/20464/files#r99494803 ?

Sacha0 commented:
Re. discussion in the review comments above, the general argument for the weaker conditions (`length(colptr) >= n + 1` rather than `length(colptr) == n + 1`, `length(rowval) >= length(colptr) - 1` rather than `length(rowval) == length(colptr) - 1` and likewise for `nzval`, and no restriction on the relative lengths of `rowval` and `nzval`) is better allocation control / predictability / reuse. Examples:

Re. `length(rowval) >= length(colptr) - 1` rather than `length(rowval) == length(colptr) - 1` (and likewise for `nzval`), suppose you want to store sparse matrices with different patterns (and particularly stored entry counts) in the same data structure at different points during execution. Under the weaker condition you can do so by allocating enough storage for all expected patterns once at the outset and reusing that storage, whereas under the stronger condition (and concomitant `resize!`ing) allocation becomes unpredictable. (This issue impacts e.g. sparse `broadcast[!]`.)

Re. `length(colptr) >= n + 1` rather than `length(colptr) == n + 1`, suppose you want to store sparse matrices with different shapes at different points during execution. Under the weaker condition you can do so by allocating enough storage for all expected shapes once at the outset and reusing that storage (constructing only a new `SparseMatrixCSC` wrapper for each new sparse matrix), whereas under the stronger condition (and concomitant `resize!`ing) allocation becomes unpredictable. (This issue impacts e.g. the applications discussed in the thread regarding immutability of `SparseMatrixCSC`.)

Re. no restrictions on the relative lengths of `rowval` and `nzval`, suppose you are operating solely on a pattern rather than a complete CSC sparse matrix. Under the weaker condition you can cut your memory footprint by almost a factor of two, whereas under the stronger condition you must carry around a potentially large dead buffer of one form or another. Though your CSC sparse matrix isn't technically valid in that scenario, that scenario is common, for example in the symbolic phases of sparse factorization or in preordering. Alternatively, suppose you want to write a preordering or factorization routine that accepts some buffers and returns a CSC sparse matrix that recycles those buffers as storage; those buffers may need to have different lengths, in which case the weaker condition grants better predictability of and control over allocation downstream. (Issues like these I encountered while working on ApproxMinimumDegree.jl.)

Of course if CHOLMOD expects any of the stronger conditions and will complain or fail if they are not satisfied, I agree wholeheartedly that checking them at CHOLMOD entry points is a great idea :).

As you note above some code in Base assumes some subset of the stronger conditions. For reasons such as the above, we might want to ween that code off those assumptions over time, and e.g. remove automatic `resize!`ing from a variety of operations. Similarly, decoupling `SparseVector` storage length from stored entry count is another thing we should strongly consider (to mirror CSC's flexibility and control). (I planned to write a little Julep on the above when things settled down a bit, but it's mostly above already :).)

Best!

andreasnoack commented:
Thanks for the explanation. Generally, I'm fine with what you suggest but we should then really take a pass through the sparse code to make sure we use the buffers accordingly. E.g. I can see that [this method](https://github.com/JuliaLang/julia/blob/7aab89d0d2846d9eefa1c1e4b849d9bdcf4fd50c/base/sparse/sparsematrix.jl#L46) makes the same assumptions as I did whereas the one right above doesn't. I'll update this PR to use your CSC definition.

Sacha0 commented:
> we should then really take a pass through the sparse code to make sure we use the buffers accordingly

Wholeheartedly agreed :). Best!

andreasnoack commented:
I've changed the tests following the conclusion above. I've also changed the `CHOLMOD` code and the `nnz` definition to follow the rules you described and I've just confirmed that it fixes #20024.

Sacha0 commented on base/sparse/cholmod.jl:
```diff
@@ -860,7 +860,18 @@ convert(::Type{Dense}, A::Sparse) = sparse_to_dense(A)
+    if length(colptr) <= n
+        throw(ArgumentError("length of colptr must be at least n + 1 = $(n + 1) but was $(length(colptr))"))
+    end
+    if colptr[n + 1] > length(rowval)
```
Should these checks be `colptr[n + 1] > length(rowval) + 1` and `colptr[n + 1] > length(nzval) + 1`? If memory serves, `colptr[n + 1]` should be one greater than the number of stored entries?

andreasnoack replied:
This constructor is zero based. I'm wondering if it should have a more alarming name for that reason.

Sacha0 replied:
> This constructor is zero based. I'm wondering if it should have a more alarming name for that reason.

That sounds great! Similarly, if some instances of `colptr` are zero based, might be worth similarly naming them appropriately? Best!

Sacha0 commented on base/sparse/cholmod.jl:
```diff
@@ -869,12 +880,12 @@ function (::Type{Sparse}){Tv<:VTypes}(m::Integer, n::Integer,
     end
 
-    o = allocate_sparse(m, n, length(nzval), iss, true, stype, Tv)
+    o = allocate_sparse(m, n, colptr[n + 1], iss, true, stype, Tv)
```
Should the third argument to `allocate_sparse` be the number of stored entries or one more than the number of stored entries?

Sacha0 commented on base/sparse/cholmod.jl:
```diff
@@ -869,12 +880,12 @@ function (::Type{Sparse}){Tv<:VTypes}(m::Integer, n::Integer,
-    unsafe_copy!(s.i, pointer(rowval), length(rowval))
-    unsafe_copy!(s.x, pointer(nzval), length(nzval))
+    unsafe_copy!(s.p, pointer(colptr), n + 1)
+    unsafe_copy!(s.i, pointer(rowval), colptr[n + 1])
```
Should these be `colptr[n + 1] - 1`, i.e. the number of stored entries?

tkelman commented on base/sparse/sparsematrix.jl:
````diff
@@ -42,7 +42,7 @@ julia> nnz(A)
 ```
 """
-nnz(S::SparseMatrixCSC) = Int(S.colptr[end]-1)
+nnz(S::SparseMatrixCSC) = Int(S.colptr[S.n + 1]-1)
````
the distinction here could use a more direct test that didn't go through cholmod-specific methods

andreasnoack replied:
I've added some tests.
